### DISCUSSION TITLE
Cannot resolve dynamic imports at runtime

### DISCUSSION THREAD

**Discussion by mwaqar666** [webpack/webpack#16566] [Q&A]
# Disussion

**What is the current behavior?**
I am using Nx. I built an app using Webpack. Now in a dependent library, there's a dynamic import statement. Webpack gives error on runtime as `Module not found` with complete path of the file that is not found. Interestingly the file is present in it's location. Basically I am finding all the files that ends with `.migration.js` using glob pattern and then importing them at runtime.

**If the current behavior is a bug, please provide the steps to reproduce.**
Here's the repro steps: 
1. Clone the repo: [https://github.com/mwaqar666/sca](https://github.com/mwaqar666/sca)
2. Switch to branch backend/backend-integration
3. Run `npx nx run sca-cmd:build` (This will build the command-line application using Webpack)
4. Run `npx nx run backend-data-access-layer:build` (This will build the project using `tsc` compiler from which files are imported dynamically)
5. Run `npx nx run sca-cmd:run:dev --verbose -- migration:up` (This simply executes the `sca-cmd` app using `node`)

**What is the expected behavior?**
The dynamic import statement should find and import the file without any error

**Other relevant information:**
webpack version:  5.75.0
Node.js version: 18.12.1
Operating System: Ubuntu 22.04.1 LTS (Jammy Jellyfish)
Additional tools: Nx v15.3.3, Typescript v4.9.4

**Comment by alexander-akait** [comment]
> Error: Config validation error: "ACCESS_TOKEN_SECRET" is required

Using `If the current behavior is a bug, please provide the steps to reproduce.`

**Reply by alexander-akait** [reply]
Also please reduce your example (if it is posssible), most often at this stage, most developers find what was the cause of the problem

**Reply by mwaqar666** [reply]
Sorry, I forgot to add `.env` file for the showcase. I have added it for now, it doesn't contain anything confidential for now. This will be good for now.

And sorry for the project size. I can't quite figure out a way to reduce the example. I tried to do it, but eventually had to add 3 libs in it

Just update the repo by regular `git pull`

**Reply by alexander-akait** [reply]
> ConnectionRefusedError [SequelizeConnectionRefusedError]: connect ECONNREFUSED 127.0.0.1:5432

Sorry, I don't have installed postgres, where is the file with dynamic import and the problem? Can I run it without extra environment, because you are saying about the `import(...)` problem, so we can debug it without it

**Reply by mwaqar666** [reply]
Actually, the `import` statement is used to import the migration files at runtime. To run the migrations, you should have database connected :(

Dynamic import is present in `migration-manager.ts` file. It is present in `sca/libs/backend/db/src/migrator/migration/migration-manager.ts`

If you have any database, you can connect to it by changing the config in `.env` file.

Just set the `DB_DIALECT` key to one of `mysql`, `mssql`, `sqlite`, `postgres`, `oracle` and ports accordingly. I have also added the db drivers, so you can update the repo by a regular `git pull`

I'm sorry for this much trouble :(

**Reply by alexander-akait** [reply]
The problem is in `import(`${path}`)`, it is not statical analizable, also after bundling your code is in another place, so `import(...)` just doesn't work, looks like this code is not written for bundling (only server usage). Why you bundle this?

**Reply by mwaqar666** [reply]
I am using Nx for the first time, so I am using the default webpack settings and didn't change much in the original config. What can I do to resolve it? Considering that the file that is going to be dynamically import could be anywhere in the `libs` folder. It can be in any project. The dynamic import files are searched using a glob pattern that ends with `.migration.js` *(js = after build)*

**Reply by alexander-akait** [reply]
I think it would be better to create an issue in `ng` and ask the developers about it. It's really hard to understand the logic here and why it was used in this way.

**Comment by jsprw** [comment]
I have encountered the same issue since upgrading to Typescript 4.9.4 with a issue that looks the same. Looking around issues I narrowed it down to these this regression:

It looks like Typescript 4.9.2+ changed the code that it emits for dynamic imports: the import path is assigned to a variable (see https://github.com/microsoft/TypeScript/issues/51554) prior to passing it to the `import()` function. Therefore Webpack does not seem to understand what to import and emits `Critical dependency: the request of a dependency is an expression` warnings. Before v4.9.2 it does not emit any warnings and it works fine.

What's emitted by Typescript 4.9.4:
```await Promise.all(filesToImport.map((e) => { var _a; return _a = `./relative/path/to/files/${e}`, Promise.resolve().then(() => tslib_1.__importStar(__webpack_require__("./path/of/app sync recursive")(_a))); }))```

What's emitted by Typescript 4.9.1-beta (and earlier) with the same project and Webpack configuration:
```await Promise.all(filesToImport.map((e) => Promise.resolve().then(() => tslib_1.__importStar(__webpack_require__("./relative/path/to/files sync recursive ^\\.\\/.*$")(`./${e}`)))))```

The solution for me is to use Typescript 4.9.1-beta for the time-being, I tried using the dynamic import magic comments but they don't seem to influence much either. 

@alexander-akait is it a known issue that Typescript 4.9.2+ emits different code for dynamic imports? Or is this something that would not influence Webpack?

**Reply by alexander-akait** [reply]
Can you provide origianl input?

**Reply by jsprw** [reply]
This is the original input: 

```
await Promise.all(filesToImport.map((e) => import(`./relative/path/to/files/${e}`)))
```

**Reply by alexander-akait** [reply]
Sorry for delat, still actual? Because I don't see something special, maybe ts bug? Maybe you can provide small reproducible test repo with such config and configuration, it requires to investigation why it happens, thank you

**Reply by jsprw** [reply]
Created an issue #16793 with a reproduction repo.
